lib/virsh_autotest_base.pm

```diff
@@ -0,0 +1,126 @@
+# VIRSH TEST MODULE BASE PACKAGE
+#
+# Copyright © 2019 SUSE LLC
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved. This file is offered as-is,
+# without any warranty.
+#
+# Summary: This is the base package for virsh test modules, for example,
+# tests/virtualization/xen/hotplugging.pm
+# tests/virt_autotest/virsh_internal_snapshot.pm
+# tests/virt_autotest/virsh_external_snapshot.pm
+# and etc.
+# Maintainer: Wayne Chen <wchen@suse.com>
+
+package virsh_autotest_base;
+
+use base "consoletest";
+use strict;
+use warnings;
+use POSIX 'strftime';
+use File::Basename;
+use Data::Dumper;
+use XML::Writer;
+use IO::File;
+use testapi;
+use utils;
+use virt_utils;
+use xen;
+
+sub run_test {
+    die('Please override this subroutine in children modules to run desired tests.');
+}
+
+sub run {
+    my ($self) = @_;
+    $self->{"start_run"} = time();
+    $self->run_test;
+    $self->{"stop_run"} = time();
+    #(caller(0))[3] helps pass calling subroutine name into called subroutine
+    $self->junit_log_provision((caller(0))[3]) if get_var("VIRT_AUTOTEST");
+}
+
+sub junit_log_provision {
+    my ($self, $runsub) = @_;
+    my $status    = eval { $runsub =~ /post_fail_hook/img ? 'FAILED' : 'PASSED' };
+    my $tc_result = $self->analyzeResult($status);
```

Comment:
Based on above 2 lines, if a job has several sub-tests, all test passed, the result is correct for subtests. But if a job has some sub-tests pass while some fail, then  it will be triggered via post_fail_hook, and all sub-test status will be FAILED? Is it reasonable?

If this PR is merged, the future new implemented virt feature tests will provide the per-guest per sub test result as a hash var for final junit log display, this function need to support it now.

Reply:
IMHO, any following improvement or enhancement should be implemented in following implementations or PRs, which is beyond the scope of this PR.

Reply:
It can be implemented in new PR, under condition that  this support is made before others implementing new tests, or every new feature test PR will be split into two , one for old style implementation, one with detailed subtest status which are quite related to previous one. This will make it effort wasting for new feature test writers. 
Can you share your plan and expected date for making this support?

Reply:
There is no need to be splitted into two. It is either based on the new one or old one. And any new feature can just focus on the "actual test" writing, no further effort need to be put into base module. I do not think I need to point it out again here, but only for demonstration, yes, it only has run_test{} subroutine. Any new feature can just pass or fail as it should. My further implementation can be very quick with a new PR or a little bit slow, it depends on upcoming activities a little bit.